internal/types/types.go

```diff
@@ -5,58 +5,68 @@ import (
 	"io"
 
 	"github.com/containers/image/v5/docker/reference"
+	"github.com/containers/image/v5/internal/blobinfocache"
 	publicTypes "github.com/containers/image/v5/types"
 )
 
+// ImageDestinationInternalOnly is the part of ImageDestinationWithOptions that is not
+// a part of publicTypes.ImageDestination.
+type ImageDestinationInternalOnly interface {
+	// PutBlobWithOptions writes contents of stream and returns data representing the result.
+	// inputInfo.Digest can be optionally provided if known; if provided, and stream is read to the end without error, the digest MUST match the stream contents.
+	// inputInfo.Size is the expected length of stream, if known.
+	// inputInfo.MediaType describes the blob format, if known.
+	// WARNING: The contents of stream are being verified on the fly.  Until stream.Read() returns io.EOF, the contents of the data SHOULD NOT be available
+	// to any other readers for download using the supplied digest.
+	// If stream.Read() at any time, ESPECIALLY at end of input, returns an error, PutBlob MUST 1) fail, and 2) delete any data stored so far.
+	PutBlobWithOptions(ctx context.Context, stream io.Reader, inputInfo publicTypes.BlobInfo, options PutBlobOptions) (publicTypes.BlobInfo, error)
+
+	// TryReusingBlobWithOptions checks whether the transport already contains, or can efficiently reuse, a blob, and if so, applies it to the current destination
+	// (e.g. if the blob is a filesystem layer, this signifies that the changes it describes need to be applied again when composing a filesystem tree).
+	// info.Digest must not be empty.
+	// If the blob has been successfully reused, returns (true, info, nil); info must contain at least a digest and size, and may
+	// include CompressionOperation and CompressionAlgorithm fields to indicate that a change to the compression type should be
+	// reflected in the manifest that will be written.
+	// If the transport can not reuse the requested blob, TryReusingBlob returns (false, {}, nil); it returns a non-nil error only on an unexpected failure.
+	TryReusingBlobWithOptions(ctx context.Context, info publicTypes.BlobInfo, options TryReusingBlobOptions) (bool, publicTypes.BlobInfo, error)
+}
+
 // ImageDestinationWithOptions is an internal extension to the ImageDestination
 // interface.
 type ImageDestinationWithOptions interface {
```

Comment:
This should probably end up just `internalTypes.ImageDestination` (or maybe `…Destination2`/`DestinationX`??).

For internal callers, it is almost attractive to alias every public type through this package, so that internal callers only ever import one `types` package and we don’t have the `internalTypes`/`publicTypes` noise.

Or maybe even `c/image/internal/image.Destination`??? (OTOH there’s `c/image/image`)

There are way too many minimally-plausible ways to slice this. Luckily this is internal, so we don’t need to get this _perfect_ immediately.

Reply:
It’s `private.ImageDestination` now.